Support for `psr/container` v2

boesing opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: default X.Y.z branch or the oldest support X.Y.z
  * Bugfix: default X.Y.z branch or the oldest support X.Y.z
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: default X.Y.z branch or the oldest support X.Y.z
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| BC Break      | no
| New Feature   | yes

### Description

<!--
Tell us about why this change is necessary:
- Are you fixing a bug or providing a failing unit test to demonstrate a bug?
  - How do you reproduce it?
  - What did you expect to happen?
  - What actually happened?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you adding documentation?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you providing a QA improvement (additional tests, CS fixes, etc.) that
  does not change behavior?
  - Explain why the changes are necessary
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you fixing a BC Break?
  - How do you reproduce it?
  - What was the previous behavior?
  - What is the current behavior?
  - TARGET THE default X.Y.z branch or the oldest support X.Y.z

- Are you adding something the library currently does not support?
  - Why should it be added?
  - What will it enable?
  - How will the code be used?
  - TARGET THE develop BRANCH

- Are you refactoring code?
  - Why do you feel the refactor is necessary?
  - What types of refactoring are you doing?
  - TARGET THE develop BRANCH
-->

Adding support for `psr/container` v2.

### TODO

- [x] bump `psr/container-implementation` in `composer.json`

boesing committed:
feature: add support for `psr/container` v2

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: remove `prophecy` usage

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: apply coding standard and ignore some psalm issues regarding `class_alias`

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing commented on src/AbstractUntypedContainerImplementation.php:
```diff
@@ -0,0 +1,33 @@
+/**
+ * @psalm-require-extends ServiceManager
+ */
+abstract class AbstractUntypedContainerImplementation implements ContainerInterface
```
I wish I could make this internal but this would trigger psalm internal method calls all over upstream applications where `ServiceManager` and `PluginManager` injections are used...

Ocramius replied:
Can we at least document intended usage then? :D

boesing replied:
done.

boesing commented on src/AbstractTypedContainerImplementation.php:
```diff
@@ -0,0 +1,35 @@
+    /**
+     * {@inheritDoc}
+     */
+    public function get(string $id)
```
should we also add `mixed` as return type here as it will be added with `psr/container` v3 anyways?

weierophinney replied:
Yes, please. It's new code; let's be explicit.

boesing replied:
Ah, can't really add return type `:mixed` since `psr/container` v2 has minimum of PHP 7.4 where `mixed` as a return value is not allowed...

weierophinney commented on composer.json:
```diff
@@ -27,20 +27,19 @@
     },
     "require": {
         "php": "~7.4.0 || ~8.0.0 || ~8.1.0",
+        "composer-plugin-api": "^2.0",
```
Why is this here?

boesing replied:
I need to know the version of the installed `psr/container` version.
Previously, we used `ocramius/package-versions` but since composer adds that feature as of composer v2, we should require users to use the composer v2 to avoid an additional dependency (such as `composer/package-versions-deprecated`).

boesing replied:
As decided in the latest TSC meeting: we avoid the usage of `composer/package-versions-deprecated` and instead use `composer-plugin-api` with at least v2.

weierophinney reviewed (approved):
Nice work! Only question I have is around the addition of the compost plugin API requirement.

boesing commented:
I want to delay this until we talked about laminas/technical-steering-committee#112

boesing committed:
qa: removed `composer/package-versions-deprecated` dev-dependency

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

Ocramius commented on composer.json:
```diff
@@ -26,21 +26,18 @@
+        "composer-plugin-api": "^2.0",
         "laminas/laminas-stdlib": "^3.2.1",
-        "psr/container": "^1.0"
+        "psr/container": "^1.1 || ^2.0"
```
Our `composer.lock` didn't get updated with this 🤔

boesing replied:
What does that actually mean?

Ocramius replied:
Means that our static analysis, CS, etc. are not checking if we're compatible with 2.x, nor that a `composer update` will bring us to 2.x

boesing replied:
Yes, we only verify one version. Thats how we do this in all packages which do support multiple majors of any kind?

Unit tests are running on v2, codestyle has nothing to do with dependency versioning so the only thing I'd say could be useful would be static analysis.

But unit tests are definitely executed for v2:
https://github.com/laminas/laminas-servicemanager/runs/7380249774?check_suite_focus=true#step:3:73

What I could do here is to add `.laminas-ci.json` additional checks to actually run psalm on v2.

Ocramius replied:
Hmm, I think the nightly lockfile upgrade will bump us to v2 right after the patch is merged, no?

boesing replied:
I have no idea if renovate will bump the lock file in this case. Do you suggest to actually have v2 within the lockfile and use a downgrade (`composer require psr/container:^1`) to run psalm for v1 instead of bumping to v2 as I do as of https://github.com/laminas/laminas-servicemanager/pull/127/commits/2d36a4ba947f0c5ed25763de9d5c7ea9c38609bd?

boesing replied:
I've bumped the lockfile to v2 and downgrade to v1 for CI.

Ocramius replied:
Awesome, indeed best approach, thanks @boesing!

boesing committed:
qa: rework abstract container implementations to get rid of unnecessary assertions

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: add psalm checks for composer v2

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: re-add platform requirement exception for PHP 8.1

`ocramius/proxy-manager` is not supporting PHP 8.1 as of now. We could replace the dependency with `friendsofphp/proxy-manager-lts` or simply ignore platform requirement for PHP 8.1. For PHP 8.2, we should remove the exception for PHP 8.1 and if there are problems with the proxy manager, we should replace it with the LTS version.

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: allow all related packages to update when requiring `psr/container` v2

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: update `psr/container` to v2 and use a downgrade to v1 for CI instead

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

Ocramius commented on src/AbstractTypedContainerImplementation.php:
```diff
@@ -0,0 +1,60 @@
+/**
+ * @psalm-require-extends ServiceManager
+ */
+abstract class AbstractTypedContainerImplementation implements ContainerInterface
```
Discussed below: `@internal` is not viable, but we should document the rough usage scope for this class

Ocramius commented on src/autoload.php:
```diff
@@ -19,3 +27,16 @@ class_alias(ContainerExceptionInterface::class, InteropContainerException::class
+
+assert(
+    $installedContainerVersion !== null,
+    'psr/container is required by `composer.json` and therefore this method should not return `null`.'
```
"this method" is a bit out of context now :D 

Anyway, not a problem, just noticed it as a tiny detail

Ocramius commented on .laminas-ci.json:
```diff
@@ -1,5 +1,15 @@
+            "job": {
+                "php": "@lowest",
+                "dependencies": "locked",
+                "command": "composer require -W psr/container:^1 && vendor/bin/psalm --stats --output-format=github --no-cache"
```
Thanks for adding a downgrade testing job! :+1:

Ocramius reviewed (approved):
LGTM - comments mostly about tiny documentation details.

I think we're good here, the only thing that comes next is rolling this out, and then observing downstream type errors propagate (related to @gsteel's recent type refinements).

Those should appear as the lockfiles upgrade in various components in the next night.

boesing committed:
docs: add some inline documentation to outline the intention of these classes

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

gsteel commented on src/AbstractTypedContainerImplementation.php:
```diff
@@ -0,0 +1,60 @@
+     * @psalm-param string|class-string $id
+     * @throws NotFoundExceptionInterface  No entry was found for **this** identifier.
+     * @throws ContainerExceptionInterface Error while retrieving the entry.
+     * @return mixed Entry.
```
Is it too presumptuous to add

```php
/**
 * @template T of object
 * @psalm-param string|class-string<T> $id
 * @psalm-return ($id is class-string ? T : mixed)
 */
```

… both for `get` and `getService`??

weierophinney replied:
I'd say we shouldn't, as the mapping is not guaranteed. As an example, you might map a class name to an alternate instance that implements the same interface, but does not extend the original class name. As such, it would be valid for the plugin manager, but the assertion that it's of the same class or inheritance tree is now invalid.

weierophinney reviewed (approved):
My only concern is the fact that the `has()` and `get()` methods now are proxies for `hasService()` and `getService()`, respectively, which means we end up with extra method calls in each case. Have you done any benchmarking to see the performance impact?

Otherwise, straight-forward (well, as straight-forward as something like this gets), and no other red flags.

boesing commented:
@weierophinney 
I did not ran any benchmarks. But since we do have phpbench in here, I can gather that information from this CI run and from the last CI.

Proxying to new methods was the easiest way to actually avoid having 100% duplication of the `ServiceManager`. Happy to hear an alternative but I have no idea how to avoid this without having to duplicate the whole service manager.

---

## Benchmarks
### v3.14.0
```plain
+-------------------------------------------------------------+-------------------------------------+-----+------+-----+----------+-----------+--------+
| benchmark                                                   | subject                             | set | revs | its | mem_peak | mode      | rstdev |
+-------------------------------------------------------------+-------------------------------------+-----+------+-----+----------+-----------+--------+
| HasNewServicesBench                                         | benchHasFactory1                    |     | 2    | 2   | 1.829mb  | 0.000μs   | ±0.00% |
| HasNewServicesBench                                         | benchHasInvokable1                  |     | 2    | 2   | 1.829mb  | 0.500μs   | ±0.00% |
| HasNewServicesBench                                         | benchHasService1                    |     | 2    | 2   | 1.829mb  | 0.000μs   | ±0.00% |
| HasNewServicesBench                                         | benchFetchFactoryAlias1             |     | 2    | 2   | 1.829mb  | 0.000μs   | ±0.00% |
| HasNewServicesBench                                         | benchHasRecursiveFactoryAlias1      |     | 2    | 2   | 1.829mb  | 0.500μs   | ±0.00% |
| HasNewServicesBench                                         | benchFetchRecursiveFactoryAlias2    |     | 2    | 2   | 1.829mb  | 1.000μs   | ±0.00% |
| HasNewServicesBench                                         | benchFetchAbstractFactoryFoo        |     | 2    | 2   | 1.829mb  | 0.500μs   | ±0.00% |
| HasNewServicesBench                                         | benchNonExistingService             |     | 2    | 2   | 1.829mb  | 0.000μs   | ±0.00% |
| FetchNewServicesBench                                       | benchFetchFactory1                  |     | 2    | 2   | 1.829mb  | 1.500μs   | ±0.00% |
| FetchNewServicesBench                                       | benchBuildFactory1                  |     | 2    | 2   | 1.829mb  | 1.500μs   | ±0.00% |
| FetchNewServicesBench                                       | benchFetchInvokable1                |     | 2    | 2   | 1.829mb  | 2.000μs   | ±0.00% |
| FetchNewServicesBench                                       | benchBuildInvokable1                |     | 2    | 2   | 1.829mb  | 2.000μs   | ±0.00% |
| FetchNewServicesBench                                       | benchFetchService1                  |     | 2    | 2   | 1.829mb  | 0.500μs   | ±0.00% |
| FetchNewServicesBench                                       | benchFetchFactoryAlias1             |     | 2    | 2   | 1.829mb  | 1.000μs   | ±0.00% |
| FetchNewServicesBench                                       | benchBuildFactoryAlias1             |     | 2    | 2   | 1.829mb  | 1.500μs   | ±0.00% |
| FetchNewServicesBench                                       | benchFetchRecursiveFactoryAlias1    |     | 2    | 2   | 1.829mb  | 2.000μs   | ±0.00% |
| FetchNewServicesBench                                       | benchBuildRecursiveFactoryAlias1    |     | 2    | 2   | 1.829mb  | 1.000μs   | ±0.00% |
| FetchNewServicesBench                                       | benchFetchRecursiveFactoryAlias2    |     | 2    | 2   | 1.829mb  | 2.000μs   | ±0.00% |
| FetchNewServicesBench                                       | benchBuildRecursiveFactoryAlias2    |     | 2    | 2   | 1.829mb  | 1.500μs   | ±0.00% |
| FetchNewServicesBench                                       | benchFetchAbstractFactoryFoo        |     | 2    | 2   | 1.829mb  | 1.500μs   | ±0.00% |
| FetchNewServicesBench                                       | benchBuildAbstractFactoryFoo        |     | 2    | 2   | 1.829mb  | 1.500μs   | ±0.00% |
| FetchNewServiceManagerBench                                 | benchFetchServiceManagerCreation    |     | 2    | 2   | 2.573mb  | 325.262μs | ±1.92% |
| SetNewServicesBench                                         | benchSetService                     |     | 2    | 2   | 1.829mb  | 0.000μs   | ±0.00% |
| SetNewServicesBench                                         | benchSetFactory                     |     | 2    | 2   | 1.829mb  | 1.000μs   | ±0.00% |
| SetNewServicesBench                                         | benchSetAlias                       |     | 2    | 2   | 1.829mb  | 2.000μs   | ±0.00% |
| SetNewServicesBench                                         | benchSetAliasOverrided              |     | 2    | 2   | 1.829mb  | 1.000μs   | ±0.00% |
| FetchNewServiceUsingReflectionAbstractFactoryAsFactoryBench | benchFetchServiceWithNoDependencies |     | 2    | 2   | 1.829mb  | 2.000μs   | ±0.00% |
| FetchNewServiceUsingReflectionAbstractFactoryAsFactoryBench | benchBuildServiceWithNoDependencies |     | 2    | 2   | 1.829mb  | 2.000μs   | ±0.00% |
| FetchNewServiceUsingReflectionAbstractFactoryAsFactoryBench | benchFetchServiceDependingOnConfig  |     | 2    | 2   | 1.829mb  | 4.000μs   | ±0.00% |
| FetchNewServiceUsingReflectionAbstractFactoryAsFactoryBench | benchBuildServiceDependingOnConfig  |     | 2    | 2   | 1.829mb  | 4.000μs   | ±0.00% |
| FetchNewServiceUsingReflectionAbstractFactoryAsFactoryBench | benchFetchServiceWithDependency     |     | 2    | 2   | 1.829mb  | 4.000μs   | ±0.00% |
| FetchNewServiceUsingReflectionAbstractFactoryAsFactoryBench | benchBuildServiceWithDependency     |     | 2    | 2   | 1.829mb  | 3.500μs   | ±0.00% |
| FetchNewServiceViaReflectionAbstractFactoryBench            | benchFetchServiceWithNoDependencies |     | 2    | 2   | 1.829mb  | 2.000μs   | ±0.00% |
| FetchNewServiceViaReflectionAbstractFactoryBench            | benchBuildServiceWithNoDependencies |     | 2    | 2   | 1.829mb  | 1.500μs   | ±0.00% |
| FetchNewServiceViaReflectionAbstractFactoryBench            | benchFetchServiceDependingOnConfig  |     | 2    | 2   | 1.829mb  | 3.000μs   | ±0.00% |
| FetchNewServiceViaReflectionAbstractFactoryBench            | benchBuildServiceDependingOnConfig  |     | 2    | 2   | 1.829mb  | 4.000μs   | ±0.00% |
| FetchNewServiceViaReflectionAbstractFactoryBench            | benchFetchServiceWithDependency     |     | 2    | 2   | 1.829mb  | 5.000μs   | ±0.00% |
| FetchNewServiceViaReflectionAbstractFactoryBench            | benchBuildServiceWithDependency     |     | 2    | 2   | 1.829mb  | 5.000μs   | ±0.00% |
| FetchNewServiceViaConfigAbstractFactoryBench                | benchFetchServiceWithNoDependencies |     | 2    | 2   | 1.829mb  | 2.500μs   | ±0.00% |
| FetchNewServiceViaConfigAbstractFactoryBench                | benchBuildServiceWithNoDependencies |     | 2    | 2   | 1.829mb  | 2.500μs   | ±0.00% |
| FetchNewServiceViaConfigAbstractFactoryBench                | benchFetchServiceDependingOnConfig  |     | 2    | 2   | 1.829mb  | 2.500μs   | ±0.00% |
| FetchNewServiceViaConfigAbstractFactoryBench                | benchBuildServiceDependingOnConfig  |     | 2    | 2   | 1.829mb  | 2.500μs   | ±0.00% |
| FetchNewServiceViaConfigAbstractFactoryBench                | benchFetchServiceWithDependency     |     | 2    | 2   | 1.829mb  | 2.500μs   | ±0.00% |
| FetchNewServiceViaConfigAbstractFactoryBench                | benchBuildServiceWithDependency     |     | 2    | 2   | 1.829mb  | 2.500μs   | ±0.00% |
| FetchCachedServicesBench                                    | benchFetchFactory1                  |     | 2    | 2   | 1.829mb  | 0.000μs   | ±0.00% |
| FetchCachedServicesBench                                    | benchFetchInvokable1                |     | 2    | 2   | 1.829mb  | 0.000μs   | ±0.00% |
| FetchCachedServicesBench                                    | benchFetchService1                  |     | 2    | 2   | 1.829mb  | 0.500μs   | ±0.00% |
| FetchCachedServicesBench                                    | benchFetchAlias1                    |     | 2    | 2   | 1.829mb  | 0.000μs   | ±0.00% |
| FetchCachedServicesBench                                    | benchFetchRecursiveAlias1           |     | 2    | 2   | 1.829mb  | 0.000μs   | ±0.00% |
| FetchCachedServicesBench                                    | benchFetchRecursiveAlias2           |     | 2    | 2   | 1.829mb  | 0.000μs   | ±0.00% |
| FetchCachedServicesBench                                    | benchFetchAbstractFactoryService    |     | 2    | 2   | 1.829mb  | 1.000μs   | ±0.00% |
| HasCachedServicesBench                                      | benchHasFactory1                    |     | 2    | 2   | 1.829mb  | 0.500μs   | ±0.00% |
| HasCachedServicesBench                                      | benchHasInvokable1                  |     | 2    | 2   | 1.829mb  | 0.000μs   | ±0.00% |
| HasCachedServicesBench                                      | benchHasService1                    |     | 2    | 2   | 1.829mb  | 0.000μs   | ±0.00% |
| HasCachedServicesBench                                      | benchHasAlias1                      |     | 2    | 2   | 1.829mb  | 0.000μs   | ±0.00% |
| HasCachedServicesBench                                      | benchHasRecursiveAlias1             |     | 2    | 2   | 1.829mb  | 0.000μs   | ±0.00% |
| HasCachedServicesBench                                      | benchHasRecursiveAlias2             |     | 2    | 2   | 1.829mb  | 0.000μs   | ±0.00% |
| HasCachedServicesBench                                      | benchHasAbstractFactoryService      |     | 2    | 2   | 1.829mb  | 0.000μs   | ±0.00% |
| HasCachedServicesBench                                      | benchNonExistingService             |     | 2    | 2   | 1.829mb  | 0.500μs   | ±0.00% |
| FetchNewServiceUsingConfigAbstractFactoryAsFactoryBench     | benchFetchServiceWithNoDependencies |     | 2    | 2   | 1.829mb  | 2.500μs   | ±0.00% |
| FetchNewServiceUsingConfigAbstractFactoryAsFactoryBench     | benchBuildServiceWithNoDependencies |     | 2    | 2   | 1.829mb  | 2.500μs   | ±0.00% |
| FetchNewServiceUsingConfigAbstractFactoryAsFactoryBench     | benchFetchServiceDependingOnConfig  |     | 2    | 2   | 1.829mb  | 2.500μs   | ±0.00% |
| FetchNewServiceUsingConfigAbstractFactoryAsFactoryBench     | benchBuildServiceDependingOnConfig  |     | 2    | 2   | 1.829mb  | 2.500μs   | ±0.00% |
| FetchNewServiceUsingConfigAbstractFactoryAsFactoryBench     | benchFetchServiceWithDependency     |     | 2    | 2   | 1.829mb  | 3.000μs   | ±0.00% |
| FetchNewServiceUsingConfigAbstractFactoryAsFactoryBench     | benchBuildServiceWithDependency     |     | 2    | 2   | 1.829mb  | 2.500μs   | ±0.00% |
+-------------------------------------------------------------+-------------------------------------+-----+------+-----+----------+-----------+--------+
```

### This branch
```plain
+-------------------------------------------------------------+-------------------------------------+-----+------+-----+----------+-----------+--------+
| benchmark                                                   | subject                             | set | revs | its | mem_peak | mode      | rstdev |
+-------------------------------------------------------------+-------------------------------------+-----+------+-----+----------+-----------+--------+
| FetchCachedServicesBench                                    | benchFetchFactory1                  |     | 2    | 2   | 1.828mb  | 0.500μs   | ±0.00% |
| FetchCachedServicesBench                                    | benchFetchInvokable1                |     | 2    | 2   | 1.828mb  | 0.500μs   | ±0.00% |
| FetchCachedServicesBench                                    | benchFetchService1                  |     | 2    | 2   | 1.828mb  | 0.000μs   | ±0.00% |
| FetchCachedServicesBench                                    | benchFetchAlias1                    |     | 2    | 2   | 1.828mb  | 0.000μs   | ±0.00% |
| FetchCachedServicesBench                                    | benchFetchRecursiveAlias1           |     | 2    | 2   | 1.828mb  | 0.000μs   | ±0.00% |
| FetchCachedServicesBench                                    | benchFetchRecursiveAlias2           |     | 2    | 2   | 1.828mb  | 0.500μs   | ±0.00% |
| FetchCachedServicesBench                                    | benchFetchAbstractFactoryService    |     | 2    | 2   | 1.828mb  | 1.500μs   | ±0.00% |
| FetchNewServiceViaReflectionAbstractFactoryBench            | benchFetchServiceWithNoDependencies |     | 2    | 2   | 1.828mb  | 2.500μs   | ±0.00% |
| FetchNewServiceViaReflectionAbstractFactoryBench            | benchBuildServiceWithNoDependencies |     | 2    | 2   | 1.828mb  | 2.500μs   | ±0.00% |
| FetchNewServiceViaReflectionAbstractFactoryBench            | benchFetchServiceDependingOnConfig  |     | 2    | 2   | 1.828mb  | 5.000μs   | ±0.00% |
| FetchNewServiceViaReflectionAbstractFactoryBench            | benchBuildServiceDependingOnConfig  |     | 2    | 2   | 1.828mb  | 4.500μs   | ±0.00% |
| FetchNewServiceViaReflectionAbstractFactoryBench            | benchFetchServiceWithDependency     |     | 2    | 2   | 1.828mb  | 5.250μs   | ±4.76% |
| FetchNewServiceViaReflectionAbstractFactoryBench            | benchBuildServiceWithDependency     |     | 2    | 2   | 1.828mb  | 5.000μs   | ±0.00% |
| SetNewServicesBench                                         | benchSetService                     |     | 2    | 2   | 1.828mb  | 0.000μs   | ±0.00% |
| SetNewServicesBench                                         | benchSetFactory                     |     | 2    | 2   | 1.828mb  | 1.000μs   | ±0.00% |
| SetNewServicesBench                                         | benchSetAlias                       |     | 2    | 2   | 1.828mb  | 2.000μs   | ±0.00% |
| SetNewServicesBench                                         | benchSetAliasOverrided              |     | 2    | 2   | 1.828mb  | 2.000μs   | ±0.00% |
| HasCachedServicesBench                                      | benchHasFactory1                    |     | 2    | 2   | 1.828mb  | 1.000μs   | ±0.00% |
| HasCachedServicesBench                                      | benchHasInvokable1                  |     | 2    | 2   | 1.828mb  | 0.500μs   | ±0.00% |
| HasCachedServicesBench                                      | benchHasService1                    |     | 2    | 2   | 1.828mb  | 0.000μs   | ±0.00% |
| HasCachedServicesBench                                      | benchHasAlias1                      |     | 2    | 2   | 1.828mb  | 0.000μs   | ±0.00% |
| HasCachedServicesBench                                      | benchHasRecursiveAlias1             |     | 2    | 2   | 1.828mb  | 0.500μs   | ±0.00% |
| HasCachedServicesBench                                      | benchHasRecursiveAlias2             |     | 2    | 2   | 1.828mb  | 0.500μs   | ±0.00% |
| HasCachedServicesBench                                      | benchHasAbstractFactoryService      |     | 2    | 2   | 1.828mb  | 0.500μs   | ±0.00% |
| HasCachedServicesBench                                      | benchNonExistingService             |     | 2    | 2   | 1.828mb  | 0.500μs   | ±0.00% |
| FetchNewServiceViaConfigAbstractFactoryBench                | benchFetchServiceWithNoDependencies |     | 2    | 2   | 1.828mb  | 2.500μs   | ±0.00% |
| FetchNewServiceViaConfigAbstractFactoryBench                | benchBuildServiceWithNoDependencies |     | 2    | 2   | 1.828mb  | 2.500μs   | ±0.00% |
| FetchNewServiceViaConfigAbstractFactoryBench                | benchFetchServiceDependingOnConfig  |     | 2    | 2   | 1.828mb  | 3.500μs   | ±0.00% |
| FetchNewServiceViaConfigAbstractFactoryBench                | benchBuildServiceDependingOnConfig  |     | 2    | 2   | 1.828mb  | 3.000μs   | ±0.00% |
| FetchNewServiceViaConfigAbstractFactoryBench                | benchFetchServiceWithDependency     |     | 2    | 2   | 1.828mb  | 3.000μs   | ±0.00% |
| FetchNewServiceViaConfigAbstractFactoryBench                | benchBuildServiceWithDependency     |     | 2    | 2   | 1.828mb  | 2.500μs   | ±0.00% |
| FetchNewServiceUsingReflectionAbstractFactoryAsFactoryBench | benchFetchServiceWithNoDependencies |     | 2    | 2   | 1.828mb  | 2.500μs   | ±0.00% |
| FetchNewServiceUsingReflectionAbstractFactoryAsFactoryBench | benchBuildServiceWithNoDependencies |     | 2    | 2   | 1.828mb  | 2.000μs   | ±0.00% |
| FetchNewServiceUsingReflectionAbstractFactoryAsFactoryBench | benchFetchServiceDependingOnConfig  |     | 2    | 2   | 1.828mb  | 3.500μs   | ±0.00% |
| FetchNewServiceUsingReflectionAbstractFactoryAsFactoryBench | benchBuildServiceDependingOnConfig  |     | 2    | 2   | 1.828mb  | 4.500μs   | ±0.00% |
| FetchNewServiceUsingReflectionAbstractFactoryAsFactoryBench | benchFetchServiceWithDependency     |     | 2    | 2   | 1.828mb  | 5.000μs   | ±0.00% |
| FetchNewServiceUsingReflectionAbstractFactoryAsFactoryBench | benchBuildServiceWithDependency     |     | 2    | 2   | 1.828mb  | 5.000μs   | ±0.00% |
| FetchNewServiceManagerBench                                 | benchFetchServiceManagerCreation    |     | 2    | 2   | 2.707mb  | 308.243μs | ±1.22% |
| HasNewServicesBench                                         | benchHasFactory1                    |     | 2    | 2   | 1.828mb  | 0.000μs   | ±0.00% |
| HasNewServicesBench                                         | benchHasInvokable1                  |     | 2    | 2   | 1.828mb  | 0.500μs   | ±0.00% |
| HasNewServicesBench                                         | benchHasService1                    |     | 2    | 2   | 1.828mb  | 0.500μs   | ±0.00% |
| HasNewServicesBench                                         | benchFetchFactoryAlias1             |     | 2    | 2   | 1.828mb  | 0.500μs   | ±0.00% |
| HasNewServicesBench                                         | benchHasRecursiveFactoryAlias1      |     | 2    | 2   | 1.828mb  | 0.500μs   | ±0.00% |
| HasNewServicesBench                                         | benchFetchRecursiveFactoryAlias2    |     | 2    | 2   | 1.828mb  | 0.500μs   | ±0.00% |
| HasNewServicesBench                                         | benchFetchAbstractFactoryFoo        |     | 2    | 2   | 1.828mb  | 0.500μs   | ±0.00% |
| HasNewServicesBench                                         | benchNonExistingService             |     | 2    | 2   | 1.828mb  | 0.500μs   | ±0.00% |
| FetchNewServicesBench                                       | benchFetchFactory1                  |     | 2    | 2   | 1.828mb  | 2.000μs   | ±0.00% |
| FetchNewServicesBench                                       | benchBuildFactory1                  |     | 2    | 2   | 1.828mb  | 1.500μs   | ±0.00% |
| FetchNewServicesBench                                       | benchFetchInvokable1                |     | 2    | 2   | 1.828mb  | 2.500μs   | ±0.00% |
| FetchNewServicesBench                                       | benchBuildInvokable1                |     | 2    | 2   | 1.828mb  | 1.500μs   | ±0.00% |
| FetchNewServicesBench                                       | benchFetchService1                  |     | 2    | 2   | 1.828mb  | 0.500μs   | ±0.00% |
| FetchNewServicesBench                                       | benchFetchFactoryAlias1             |     | 2    | 2   | 1.828mb  | 1.500μs   | ±0.00% |
| FetchNewServicesBench                                       | benchBuildFactoryAlias1             |     | 2    | 2   | 1.828mb  | 2.000μs   | ±0.00% |
| FetchNewServicesBench                                       | benchFetchRecursiveFactoryAlias1    |     | 2    | 2   | 1.828mb  | 2.000μs   | ±0.00% |
| FetchNewServicesBench                                       | benchBuildRecursiveFactoryAlias1    |     | 2    | 2   | 1.828mb  | 1.000μs   | ±0.00% |
| FetchNewServicesBench                                       | benchFetchRecursiveFactoryAlias2    |     | 2    | 2   | 1.828mb  | 2.000μs   | ±0.00% |
| FetchNewServicesBench                                       | benchBuildRecursiveFactoryAlias2    |     | 2    | 2   | 1.828mb  | 1.500μs   | ±0.00% |
| FetchNewServicesBench                                       | benchFetchAbstractFactoryFoo        |     | 2    | 2   | 1.828mb  | 2.000μs   | ±0.00% |
| FetchNewServicesBench                                       | benchBuildAbstractFactoryFoo        |     | 2    | 2   | 1.828mb  | 1.000μs   | ±0.00% |
| FetchNewServiceUsingConfigAbstractFactoryAsFactoryBench     | benchFetchServiceWithNoDependencies |     | 2    | 2   | 1.828mb  | 2.500μs   | ±0.00% |
| FetchNewServiceUsingConfigAbstractFactoryAsFactoryBench     | benchBuildServiceWithNoDependencies |     | 2    | 2   | 1.828mb  | 2.500μs   | ±0.00% |
| FetchNewServiceUsingConfigAbstractFactoryAsFactoryBench     | benchFetchServiceDependingOnConfig  |     | 2    | 2   | 1.828mb  | 3.000μs   | ±0.00% |
| FetchNewServiceUsingConfigAbstractFactoryAsFactoryBench     | benchBuildServiceDependingOnConfig  |     | 2    | 2   | 1.828mb  | 3.000μs   | ±0.00% |
| FetchNewServiceUsingConfigAbstractFactoryAsFactoryBench     | benchFetchServiceWithDependency     |     | 2    | 2   | 1.828mb  | 3.500μs   | ±0.00% |
| FetchNewServiceUsingConfigAbstractFactoryAsFactoryBench     | benchBuildServiceWithDependency     |     | 2    | 2   | 1.828mb  | 2.500μs   | ±0.00% |
+-------------------------------------------------------------+-------------------------------------+-----+------+-----+----------+-----------+--------+
```

@Ocramius since you have massive experience with benchmark stuff, could you share the setup you use to compare two benchmarks? I know I can actually store the result and tell phpbench to compare two results but I am pretty sure that the biggest problem would be to verify that the environment between both runs is mostly identical.

Ocramius commented:
@boesing I can only suggest using higher `--revs`, higher `--iterations`, a `--warmup=2` (to prevent any autoloading-related issues) and a `--retry-threshold` that forces a re-run.

Then make sure it's a machine where literally nothing else runs (I use my local dev environment, but with everything else closed, including editor).

I think we will take a very minor performance hit here, and then remove it once we move to `psr/container:^2` **only**.

weierophinney commented:
> I think we will take a very minor performance hit here

That's my assumption as well, but just want some verification before we turn this on. If it turns out to be a large impact, then I'd likely recommend we maintain two majors for a set duration of time; otherwise, carry on with this until we want to drop v1 support.

lcobucci commented:
@boesing you can create a baseline with the previous implementation and do diffs against it: https://phpbench.readthedocs.io/en/latest/guides/regression-testing.html

Ocramius commented:
BTW, I think this is the least impacting approach taken: we should really just take whichever impact this comes with, and then move on to `psr/container:^2` in a new minor release (which can literally be released in a week from now, giving us a `^1 || ^2`-compatible transition release meanwhile)

weierophinney commented:
> which can literally be released in a week from now

I think a v1-compatible version will need to be a bit longer-lived. It's going to take a while before we've updated our own repositories to hint against v2, much less third-party packages in our ecosystem. Even if we release a new major right away, we'll need to likely extend support for the v3 release to give time for people to update and adapt.

boesing commented:
I've executed benchmarks locally:

```plain
+-------------------------------------------------------------+-------------------------------------+-----+------+-----+----------------+------------------+-------------------------------+
| benchmark                                                   | subject                             | set | revs | its | mem_peak       | mode             | rstdev                        |
+-------------------------------------------------------------+-------------------------------------+-----+------+-----+----------------+------------------+-------------------------------+
| SetNewServicesBench                                         | benchSetService                     |     | 10   | 10  | 1.849mb -0.05% | 0.300μs 0.00%    | ±0.00% 0.00%                  |
| SetNewServicesBench                                         | benchSetFactory                     |     | 10   | 10  | 1.849mb -0.05% | 0.600μs 0.00%    | ±0.00% 0.00%                  |
| SetNewServicesBench                                         | benchSetAlias                       |     | 10   | 10  | 1.849mb -0.05% | 1.000μs 0.00%    | ±0.00% 0.00%                  |
| SetNewServicesBench                                         | benchSetAliasOverrided              |     | 10   | 10  | 1.849mb -0.05% | 1.000μs 0.00%    | ±0.00% 0.00%                  |
| FetchNewServiceUsingReflectionAbstractFactoryAsFactoryBench | benchFetchServiceWithNoDependencies |     | 10   | 10  | 1.849mb -0.05% | 1.200μs +9.09%   | ±0.00% -8.33%                 |
| FetchNewServiceUsingReflectionAbstractFactoryAsFactoryBench | benchBuildServiceWithNoDependencies |     | 10   | 10  | 1.849mb -0.05% | 1.000μs 0.00%    | ±0.00% 0.00%                  |
| FetchNewServiceUsingReflectionAbstractFactoryAsFactoryBench | benchFetchServiceDependingOnConfig  |     | 10   | 10  | 1.849mb -0.05% | 2.203μs +10.17%  | ±3.17% +112.22%               |
| FetchNewServiceUsingReflectionAbstractFactoryAsFactoryBench | benchBuildServiceDependingOnConfig  |     | 10   | 10  | 1.849mb -0.05% | 2.000μs +0.02%   | ±2.26% +49.74%                |
| FetchNewServiceUsingReflectionAbstractFactoryAsFactoryBench | benchFetchServiceWithDependency     |     | 10   | 10  | 1.849mb -0.05% | 2.501μs +1.10%   | ±1.93% -44.04%                |
| FetchNewServiceUsingReflectionAbstractFactoryAsFactoryBench | benchBuildServiceWithDependency     |     | 10   | 10  | 1.849mb -0.05% | 2.499μs +5.31%   | ±1.99% -44.62%                |
| FetchNewServiceUsingConfigAbstractFactoryAsFactoryBench     | benchFetchServiceWithNoDependencies |     | 10   | 10  | 1.849mb -0.05% | 1.500μs +0.04%   | ±0.00% -100.00%               |
| FetchNewServiceUsingConfigAbstractFactoryAsFactoryBench     | benchBuildServiceWithNoDependencies |     | 10   | 10  | 1.849mb -0.05% | 1.499μs +7.10%   | ±3.36% +21156341821897000.00% |
| FetchNewServiceUsingConfigAbstractFactoryAsFactoryBench     | benchFetchServiceDependingOnConfig  |     | 10   | 10  | 1.849mb -0.05% | 1.701μs +6.34%   | ±2.86% -2.11%                 |
| FetchNewServiceUsingConfigAbstractFactoryAsFactoryBench     | benchBuildServiceDependingOnConfig  |     | 10   | 10  | 1.849mb -0.05% | 1.600μs +6.68%   | ±2.81% +0.00%                 |
| FetchNewServiceUsingConfigAbstractFactoryAsFactoryBench     | benchFetchServiceWithDependency     |     | 10   | 10  | 1.849mb -0.05% | 1.800μs +16.13%  | ±0.00% -100.00%               |
| FetchNewServiceUsingConfigAbstractFactoryAsFactoryBench     | benchBuildServiceWithDependency     |     | 10   | 10  | 1.849mb -0.05% | 1.601μs +6.71%   | ±2.99% +0.00%                 |
| FetchNewServiceManagerBench                                 | benchFetchServiceManagerCreation    |     | 10   | 10  | 3.917mb +3.48% | 297.002μs -0.05% | ±1.94% +15.25%                |
| HasNewServicesBench                                         | benchHasFactory1                    |     | 10   | 10  | 1.849mb -0.05% | 0.300μs +50.00%  | ±0.00% +33.33%                |
| HasNewServicesBench                                         | benchHasInvokable1                  |     | 10   | 10  | 1.849mb -0.05% | 0.400μs +33.33%  | ±0.00% -25.00%                |
| HasNewServicesBench                                         | benchHasService1                    |     | 10   | 10  | 1.849mb -0.05% | 0.200μs 0.00%    | ±0.00% 0.00%                  |
| HasNewServicesBench                                         | benchFetchFactoryAlias1             |     | 10   | 10  | 1.849mb -0.05% | 0.400μs +33.33%  | ±0.00% -25.00%                |
| HasNewServicesBench                                         | benchHasRecursiveFactoryAlias1      |     | 10   | 10  | 1.849mb -0.05% | 0.400μs 0.00%    | ±0.00% 0.00%                  |
| HasNewServicesBench                                         | benchFetchRecursiveFactoryAlias2    |     | 10   | 10  | 1.849mb -0.05% | 0.400μs +33.33%  | ±0.00% -25.00%                |
| HasNewServicesBench                                         | benchFetchAbstractFactoryFoo        |     | 10   | 10  | 1.849mb -0.05% | 0.400μs 0.00%    | ±0.00% 0.00%                  |
| HasNewServicesBench                                         | benchNonExistingService             |     | 10   | 10  | 1.849mb -0.05% | 0.400μs 0.00%    | ±0.00% 0.00%                  |
| FetchNewServiceViaConfigAbstractFactoryBench                | benchFetchServiceWithNoDependencies |     | 10   | 10  | 1.849mb -0.05% | 1.500μs +0.04%   | ±0.00% -100.00%               |
| FetchNewServiceViaConfigAbstractFactoryBench                | benchBuildServiceWithNoDependencies |     | 10   | 10  | 1.849mb -0.05% | 1.399μs +7.65%   | ±3.60% +21089672677500000.00% |
| FetchNewServiceViaConfigAbstractFactoryBench                | benchFetchServiceDependingOnConfig  |     | 10   | 10  | 1.849mb -0.05% | 1.601μs +0.07%   | ±3.03% +21835634556948000.00% |
| FetchNewServiceViaConfigAbstractFactoryBench                | benchBuildServiceDependingOnConfig  |     | 10   | 10  | 1.849mb -0.05% | 1.501μs +0.04%   | ±3.18% +0.00%                 |
| FetchNewServiceViaConfigAbstractFactoryBench                | benchFetchServiceWithDependency     |     | 10   | 10  | 1.849mb -0.05% | 1.600μs +6.64%   | ±2.81% -11.62%                |
| FetchNewServiceViaConfigAbstractFactoryBench                | benchBuildServiceWithDependency     |     | 10   | 10  | 1.849mb -0.05% | 1.500μs 0.00%    | ±0.00% 0.00%                  |
| FetchNewServicesBench                                       | benchFetchFactory1                  |     | 10   | 10  | 1.849mb -0.05% | 1.000μs 0.00%    | ±0.00% 0.00%                  |
| FetchNewServicesBench                                       | benchBuildFactory1                  |     | 10   | 10  | 1.849mb -0.05% | 0.800μs -11.11%  | ±0.00% +12.50%                |
| FetchNewServicesBench                                       | benchFetchInvokable1                |     | 10   | 10  | 1.849mb -0.05% | 1.100μs +10.00%  | ±0.00% +0.00%                 |
| FetchNewServicesBench                                       | benchBuildInvokable1                |     | 10   | 10  | 1.849mb -0.05% | 0.900μs 0.00%    | ±0.00% 0.00%                  |
| FetchNewServicesBench                                       | benchFetchService1                  |     | 10   | 10  | 1.849mb -0.05% | 0.200μs +100.00% | ±0.00% 0.00%                  |
| FetchNewServicesBench                                       | benchFetchFactoryAlias1             |     | 10   | 10  | 1.849mb -0.05% | 0.800μs 0.00%    | ±0.00% 0.00%                  |
| FetchNewServicesBench                                       | benchBuildFactoryAlias1             |     | 10   | 10  | 1.849mb -0.05% | 0.800μs -11.11%  | ±0.00% +12.50%                |
| FetchNewServicesBench                                       | benchFetchRecursiveFactoryAlias1    |     | 10   | 10  | 1.849mb -0.05% | 0.800μs 0.00%    | ±0.00% 0.00%                  |
| FetchNewServicesBench                                       | benchBuildRecursiveFactoryAlias1    |     | 10   | 10  | 1.849mb -0.05% | 0.900μs 0.00%    | ±0.00% 0.00%                  |
| FetchNewServicesBench                                       | benchFetchRecursiveFactoryAlias2    |     | 10   | 10  | 1.849mb -0.05% | 0.800μs 0.00%    | ±0.00% 0.00%                  |
| FetchNewServicesBench                                       | benchBuildRecursiveFactoryAlias2    |     | 10   | 10  | 1.849mb -0.05% | 0.800μs -11.11%  | ±0.00% +12.50%                |
| FetchNewServicesBench                                       | benchFetchAbstractFactoryFoo        |     | 10   | 10  | 1.849mb -0.05% | 0.800μs 0.00%    | ±0.00% 0.00%                  |
| FetchNewServicesBench                                       | benchBuildAbstractFactoryFoo        |     | 10   | 10  | 1.849mb -0.05% | 0.600μs 0.00%    | ±0.00% 0.00%                  |
| FetchNewServiceViaReflectionAbstractFactoryBench            | benchFetchServiceWithNoDependencies |     | 10   | 10  | 1.849mb -0.05% | 1.200μs 0.00%    | ±0.00% 0.00%                  |
| FetchNewServiceViaReflectionAbstractFactoryBench            | benchBuildServiceWithNoDependencies |     | 10   | 10  | 1.849mb -0.05% | 1.100μs +10.00%  | ±0.00% +0.00%                 |
| FetchNewServiceViaReflectionAbstractFactoryBench            | benchFetchServiceDependingOnConfig  |     | 10   | 10  | 1.849mb -0.05% | 2.300μs +0.81%   | ±2.33% -42.66%                |
| FetchNewServiceViaReflectionAbstractFactoryBench            | benchBuildServiceDependingOnConfig  |     | 10   | 10  | 1.849mb -0.05% | 2.101μs +5.02%   | ±2.29% +1.41%                 |
| FetchNewServiceViaReflectionAbstractFactoryBench            | benchFetchServiceWithDependency     |     | 10   | 10  | 1.849mb -0.05% | 2.627μs +5.06%   | ±3.09% +158.36%               |
| FetchNewServiceViaReflectionAbstractFactoryBench            | benchBuildServiceWithDependency     |     | 10   | 10  | 1.849mb -0.05% | 2.500μs +4.14%   | ±1.20% -40.47%                |
| FetchCachedServicesBench                                    | benchFetchFactory1                  |     | 10   | 10  | 1.849mb -0.05% | 0.200μs 0.00%    | ±0.00% 0.00%                  |
| FetchCachedServicesBench                                    | benchFetchInvokable1                |     | 10   | 10  | 1.849mb -0.05% | 0.400μs 0.00%    | ±0.00% 0.00%                  |
| FetchCachedServicesBench                                    | benchFetchService1                  |     | 10   | 10  | 1.849mb -0.05% | 0.200μs 0.00%    | ±0.00% 0.00%                  |
| FetchCachedServicesBench                                    | benchFetchAlias1                    |     | 10   | 10  | 1.849mb -0.05% | 0.200μs 0.00%    | ±0.00% 0.00%                  |
| FetchCachedServicesBench                                    | benchFetchRecursiveAlias1           |     | 10   | 10  | 1.849mb -0.05% | 0.200μs 0.00%    | ±0.00% 0.00%                  |
| FetchCachedServicesBench                                    | benchFetchRecursiveAlias2           |     | 10   | 10  | 1.849mb -0.05% | 0.200μs 0.00%    | ±0.00% 0.00%                  |
| FetchCachedServicesBench                                    | benchFetchAbstractFactoryService    |     | 10   | 10  | 1.849mb -0.05% | 0.800μs +14.29%  | ±0.00% -12.50%                |
| HasCachedServicesBench                                      | benchHasFactory1                    |     | 10   | 10  | 1.849mb -0.05% | 0.200μs 0.00%    | ±0.00% 0.00%                  |
| HasCachedServicesBench                                      | benchHasInvokable1                  |     | 10   | 10  | 1.849mb -0.05% | 0.400μs +33.33%  | ±0.00% -25.00%                |
| HasCachedServicesBench                                      | benchHasService1                    |     | 10   | 10  | 1.849mb -0.05% | 0.300μs +50.00%  | ±0.00% +33.33%                |
| HasCachedServicesBench                                      | benchHasAlias1                      |     | 10   | 10  | 1.849mb -0.05% | 0.200μs 0.00%    | ±0.00% 0.00%                  |
| HasCachedServicesBench                                      | benchHasRecursiveAlias1             |     | 10   | 10  | 1.849mb -0.05% | 0.200μs 0.00%    | ±0.00% 0.00%                  |
| HasCachedServicesBench                                      | benchHasRecursiveAlias2             |     | 10   | 10  | 1.849mb -0.05% | 0.200μs 0.00%    | ±0.00% 0.00%                  |
| HasCachedServicesBench                                      | benchHasAbstractFactoryService      |     | 10   | 10  | 1.849mb -0.05% | 0.400μs 0.00%    | ±0.00% 0.00%                  |
| HasCachedServicesBench                                      | benchNonExistingService             |     | 10   | 10  | 1.849mb -0.05% | 0.500μs +25.00%  | ±0.00% -100.00%               |
+-------------------------------------------------------------+-------------------------------------+-----+------+-----+----------------+------------------+-------------------------------+
```

Commands used:

```
max@maxbookpro [~/git/laminas/laminas-servicemanager]  (3.14.0) via 🐘 v7.4.28 
→ php -n vendor/bin/phpbench run --revs=10 --iterations=10 --warmup=2 --retry-threshold=5 --tag=stable

max@maxbookpro [~/git/laminas/laminas-servicemanager]  feature/psr-container-2.0= via 🐘 v7.4.28 
→ php -n vendor/bin/phpbench run --revs=10 --iterations=10 --warmup=2 --retry-threshold=5 --ref=stable --report=aggregate
```

I have absolutely no idea if thats good or bad...

Ocramius commented:
@boesing it's well within tolerance IMO.

Biggest slowdowns are in `benchHasFactory1`, which is because that method effectively is almost empty :D 

:ship: here.

Ocramius commented:
> bump `psr/container-implementation` in `composer.json`

Is this part to be considered "done"?

boesing commented:
~No, I have to bump that. Forgot it in the initial PR.~

@Ocramius done.

boesing committed:
qa: bump `psr/container-implementation` to reflect implementation details

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

boesing committed:
qa: require at least v2.0.2 of `psr/container` to avoid installation of lower v2 versions

Signed-off-by: Maximilian Bösing <2189546+boesing@users.noreply.github.com>

Ocramius commented:
Awesome, thanks!

Ocramius commented:
Interesting side-effect here: breaks old plugin manager setup, such as `Laminas\Feed\Reader\ExtensionManagerInterface`.

https://github.com/laminas/laminas-feed/runs/7405125679?check_suite_focus=true

It's such a mess overall @_@ 

Oh well, potentially BC fixing patches needed all over the place, and forcing the strict typing, ignoring inheritance BC breaks in a lot of locations.

boesing commented:
I guess that is because these packages do not explicitly require `psr/container` which might be the main problem?

Ocramius commented:
That too - they are kinda "in flux" when depending on just `laminas/laminas-servicemanager` :thinking:

boesing commented:
Exactly. Inheritance is not a feature, its a bug 😂

weierophinney commented:
The fees package is an outlier here - I did a bunch of work a number of years ago to make it work "standalone" due to requests from projects like Drupal. That means it has straight up vanilla psr-11 implementations, as well as variants that build on the SM.
